src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp

```diff
@@ -0,0 +1,181 @@
+        if (input_descriptor->m_body_parameter_index == parameter_index)
+            return ti->input(input_descriptor->m_input_index);
+    OPENVINO_UNREACHABLE("LSTMStatesBroadcast failed to get outer input of TI by its inner Parameter. TI ",
+                         ti,
+                         " Parameter ",
+                         parameter);
+}
+
+std::shared_ptr<ov::Node> deduce_outer_source_of_batch_for_inner_lstm_cell(
+    const std::shared_ptr<ov::opset9::TensorIterator>& ti,
+    const std::shared_ptr<ov::opset9::LSTMCell>& lstm_cell) {
+    const auto& body = ti->get_body();
+    OPENVINO_ASSERT(body != nullptr, "TI returns invalid body graph ", ti);
+
+    std::map<ov::opset9::Parameter*, ov::PartialShape> original_shapes;
+    size_t label = 1;
+
+    // mark all input dimensions with labels and making them dynamic, keeping original shapes
+    for (auto& parameter : body->get_parameters()) {
+        auto pshape = parameter->get_partial_shape();
+        original_shapes[parameter.get()] = pshape;
+        if (pshape.rank().is_dynamic())
+            continue;
+        for (ngraph::Dimension& n : pshape) {
+            OPENVINO_ASSERT(ov::DimensionTracker::get_label(n) == 0,
+                            "LSTMStatesBroadcast encountered TI with previously tracked dimensions");
+            n = ov::Dimension::dynamic();
+            ov::DimensionTracker::set_label(n, label++);
+        }
+        parameter->set_partial_shape(pshape);
+    }
+
+    // propagate labels through TI body
+    body->validate_nodes_and_infer_types();
+    // if lstm first input has undefined rank or if tracked label is zero -- we failed to track batch dimension
+    // returning body to initial state
+    if (lstm_cell->get_input_partial_shape(0).rank().is_dynamic() ||
+        ov::DimensionTracker::get_label(lstm_cell->get_input_partial_shape(0)[0]) == 0) {
+        for (auto& item : original_shapes)
+            item.first->set_partial_shape(item.second);
+        body->validate_nodes_and_infer_types();
+        return nullptr;
+    }
+
+    // batch label was tracked -- finding parameter that delivered it
+    std::shared_ptr<ov::opset9::Parameter> batch_delivering_parameter;
+    size_t index_of_batch_dim = 0;
+
+    size_t batch_label = ov::DimensionTracker::get_label(lstm_cell->get_input_partial_shape(0)[0]);
+    for (auto& parameter : body->get_parameters()) {
+        auto pshape = parameter->get_partial_shape();
+        if (pshape.rank().is_dynamic())
+            continue;
+        for (size_t i = 0; i < pshape.size(); ++i) {
+            if (ov::DimensionTracker::get_label(pshape[i]) == batch_label) {
+                batch_delivering_parameter = parameter;
+                index_of_batch_dim = i;
+            }
+        }
+    }
+    for (auto& item : original_shapes)
+        item.first->set_partial_shape(item.second);
+    body->validate_nodes_and_infer_types();
+
+    if (batch_delivering_parameter == nullptr)
+        return nullptr;
```

Comment:
As `nullptr` is returned should we do action in above loop and `body->validate_nodes_and_infer_types();`?
If yes maybe add some comment why is done and `nullptr` returned?

Reply:
No we shouldn't. We must return function to it's original state.